File: 2020/05_binary_boarding/aoc05.jl
# Binary Boarding
#
# Advent of Code 2020, day 5
# Solution by Geir Arne Hjelle, 2020-12-05

module AOC05

using Pipe

PASS2SEATID = Dict('F' => '0', 'B' => '1', 'L' => '0', 'R' => '1')


# Parse a boarding pass into a binary seat ID
function parse_boarding_pass(boarding_pass)
    @pipe boarding_pass |> map(c -> PASS2SEATID[c], _) |> parse(Int16, _, base=2)
end


# Find missing seat IDs in a list of IDs
function find_missing(seat_ids)
    @pipe Set(minimum(seat_ids):maximum(seat_ids)) |> setdiff(_, Set(seat_ids))
end


# Solve the problem for one file
function solve(input)
    # Parse input
    seat_ids = @pipe input |> split .|> parse_boarding_pass

    # Part 1
    part_1 = seat_ids |> maximum

    # Part 2
    part_2 = @pipe seat_ids |> find_missing |> join(_, ", ")

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: 2020/05_binary_boarding/aoc05.py
"""Binary Boarding

Advent of Code 2020, day 5
Solution by Geir Arne Hjelle, 2020-12-05
"""
import pathlib
import sys

debug = print if "--debug" in sys.argv else lambda *_: None

PASS2BIN = {"F": "0", "B": "1", "L": "0", "R": "1"}


def decode_seat_id(boarding_pass):
    """Decode a boarding pass string as a seat id"""
    return int("".join(PASS2BIN[c] for c in boarding_pass), 2)


def find_missing(seat_ids):
    """Find the missing seat_ids in a list of nearby seat IDs"""
    all_ids = set(range(min(seat_ids), max(seat_ids) + 1))
    return [str(id) for id in all_ids - set(seat_ids)]


def main(args):
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        print(f"\n{file_path}:")
        seat_ids = [decode_seat_id(bp.strip()) for bp in file_path.open()]

        # Part 1
        print(f"The highest seat ID is {max(seat_ids)}")

        # Part 2
        print(f"The missing seat ID is {', '.join(find_missing(seat_ids))}")


if __name__ == "__main__":
    main(sys.argv[1:])
